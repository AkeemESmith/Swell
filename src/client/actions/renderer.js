const { ipcRenderer } = require('electron');

const ReqResCtrl = require('./src/client/controllers/connectionController')

ipcRenderer.on('openAllSelected', () => {
  console.log('here')
  ReqResCtrl.openAllSelectedReqRes()
})