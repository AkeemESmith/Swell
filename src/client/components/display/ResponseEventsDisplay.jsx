import React from "react";
import JSONPretty from "react-json-pretty";
import createDOMPurify from "dompurify";
import SSERow from "./SSERow.jsx";

const ResponseEventsDisplay = (props) => {
  const { events, headers } = props.content.response;
  const displayContents = [];
  const className = props.content.connection === 'error' ? '__json-pretty__error' : '__json-pretty__';

  // If it's an SSE, render event rows
  // console.log('response is : ', response, 'headers is : ', headers)
  if (
    headers &&
    headers["content-type"] &&
    headers["content-type"] === "text/event-stream"
  ) {
    events.forEach((cur, idx) => {
      displayContents.push(<SSERow key={idx} content={cur} />);
    });
  }
  // if the response content-type, purify and render html
  else if (
    headers &&
    headers["content-type"] &&
    headers["content-type"].includes("text/html")
  ) {
    // console.log("headers:content-type ->", headers["content-type"]);
    // console.log("events0 -> ", events[0]);
    displayContents.push(
      <div
        className="okay"
        key="http2_html_content"
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{
          __html: createDOMPurify.sanitize(events[0]),
        }}
      />
    );
  } else if (events && events.length > 1) {
    if (events) {
      let resEvents = "";
      let eventJSON;
      for (const event of events) {
        eventJSON = JSON.stringify(event, null, 4);
        resEvents = `${resEvents}
${eventJSON}`;
      }
      displayContents.push(
        <div className="json-response" key="jsonresponsediv">
          <JSONPretty
            data={resEvents}
            // onJSONPrettyError={e => console.error(e)}
            space="4"
            className={className}//theme={{
            //   main: 'line-height:1.3; color: midnightblue; background:#RRGGBB; overflow:auto;',
            //   key: 'color:#0089D0;', // bluetwo
            //   string: 'color:#15B78F;',// greenone
            //   value: 'color:#fd971f;', // a nice orange
            //   boolean: 'color:#E00198;', // gqlpink
            // }}
          />
        </div>
      );
    }
  }

  // Otherwise, render a single display
  else if (events) {
    displayContents.push(
      <div className="json-response" key="jsonresponsediv">
        <JSONPretty
          data={events[0]}
          // onJSONPrettyError={e => console.error(e)}
          space="4"
          className={className}
          // theme={{
          //   main:
          //     "line-height:1.3; color: midnightblue; background:#RRGGBB; overflow:auto;",
          //   key: "color:#0089D0;", // bluetwo
          //   string: "color:#15B78F;", // greenone
          //   value: "color:#fd971f;", // a nice orange
          //   boolean: "color:#E00198;", // gqlpink
          // }}
        />
      </div>
    );
  }

  return <div className="tab_content-response">{displayContents}</div>;
};

export default ResponseEventsDisplay;
