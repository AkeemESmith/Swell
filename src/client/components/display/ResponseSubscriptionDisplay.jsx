import React from 'react';
import JSONPretty from 'react-json-pretty';
import gql from 'graphql-tag';
import { ApolloProvider, Subscription } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { WebSocketLink } from 'apollo-link-ws';


const ResponseSubscriptionDisplay = ({ content, reqResUpdate }) => {
  let { body, bodyVariables } = content.request;
  if (bodyVariables === '') bodyVariables = null
  const uri = /wss?:\/\//.test(content.protocol) ? content.url : content.url.replace(content.protocol, 'ws://');

  const link = new WebSocketLink({
    uri,
    options: { reconnect: true }
  });

  const client = new ApolloClient({
    link,
    cache: new InMemoryCache(),
  });

  const theme = {
    main: 'line-height:1.3; color: midnightblue; background:#RRGGBB; overflow:auto;',
    key: 'color:#0089D0;', // bluetwo
    string: 'color:#15B78F;',// greenone
    value: 'color:#fd971f;', // a nice orange
    boolean: 'color:#E00198;', // gqlpink
  }

  return (
    <ApolloProvider client={client} >
      <div className="tab_content-response">
        <div className="json-response" key="jsonresponsediv">
          {content.connection === 'closed' && <JSONPretty data={content.response.events[0]} space="4" theme={theme} />}
          {content.connection === 'open' && <Subscription subscription={gql`${body}`} variables={JSON.parse(bodyVariables)}>
            {({ loading, data }) => {
              if (loading && !content.response.events[0]) return 'Listening for new data';
              if (loading && content.response.events[0]) return <JSONPretty data={content.response.events[0]} space="4" theme={theme} />
              content.response.events[0] = data;
              reqResUpdate(content);
              return <JSONPretty data={data} space="4" theme={theme} />
            }}
          </Subscription>}
        </div>
      </div>
    </ApolloProvider >
  );
}

export default ResponseSubscriptionDisplay;
