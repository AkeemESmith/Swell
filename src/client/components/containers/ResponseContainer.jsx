import React, { Component } from 'react';
import { connect } from 'react-redux';

import * as actions from '../../actions/actions';
import ResponseSSE from '../display/ResponseSSE.jsx';

const mapStateToProps = store => ({
 
});

const mapDispatchToProps = dispatch => ({

});

class ResponseContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
  
  componentDidMount () {
    if(this.state.responseDisplay !== this.props.connectionType){
      this.setState({
        responseDisplay : this.props.connectionType
      })
    } 
  }

  componentDidUpdate () {
    if(this.state.responseDisplay !== this.props.connectionType){
      this.setState({
        responseDisplay : this.props.connectionType
      })
    }
  }

  render() {
    let responseContents;
    switch (this.state.responseDisplay) {
      case 'SSE' : {
        responseContents = <ResponseSSE content={this.props.content}/>
        break;
      }
    }

    let headersArr = [];
    let index = 0;
    if(this.props.content.headers){
      for (let header in this.props.content.headers) {
        headersArr.push(<div style={{'display' : 'flex'}} key={index}>
          <div style={{'width' : '50%'}}>{header}</div>
          <div style={{'width' : '50%'}}>{this.props.content.headers[header]}</div>
        </div>);
        index++;
      }
    }
   
    return(
      <div style={{'border' : '1px solid black', 'margin' : '3px', 'display' : 'flex', 'flexDirection' : 'column'}}>
        ResponseContainer
        {headersArr}
        {responseContents}
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ResponseContainer);