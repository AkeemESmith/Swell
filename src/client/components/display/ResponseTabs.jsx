import React, { Component } from "react";
import { connect } from "react-redux";
import { LightAsync as SyntaxHighlighter } from "react-syntax-highlighter";
import { solarizedDark } from 'react-syntax-highlighter/styles/hljs';
import pretty from 'pretty';

import * as actions from "../../actions/actions";

import Tab from "./Tab.jsx";
import SSERow from "./SSERow.jsx";
import ResponsePlain from "./ResponsePlain.jsx";

const mapStateToProps = store => ({
  store: store
});

const mapDispatchToProps = dispatch => ({});

class ResponseTabs extends Component {
  constructor(props) {
    super(props);
    this.state = {
      openTabs: "",
    };
    this.handleTabSelect = this.handleTabSelect.bind(this);
  }

  handleTabSelect(val) {
    // console.log("handleTabSelect");
    switch (val) {
      case "Cookies":
        this.setState({
          openTabs: val,
        });
        break;
      case "Headers":
        this.setState({
          openTabs: val,
        });
        break;
      case "Events":
        this.setState({
          openTabs: val,
        });
        break;
      default:
        // console.log(`There was an error with ${val}`);
    }
  }

  componentDidMount() {
    this.handleTabSelect("Events");
  }

  render() {
    let events = "Events";
    // let cookies = "Cookies";
    let headers = "Headers";
    let tabContentShown;
    let tabContentShownEvents = [];

    if (this.state.openTabs === "Events" && this.props.responseContent.events) {
        tabContentShown = this.props.responseContent.events;
        tabContentShown.forEach((cur, idx) => {
            if (this.props.store.business.reqResArray[0].connectionType === 'SSE') {
                console.log('TESTING 1')
                tabContentShownEvents.push(<SSERow key={idx} content={cur} />);
            } else if (this.props.store.business.reqResArray[0].connectionType === 'plain') {
                console.log('TESTING 2')
                tabContentShownEvents.push(
                    <SyntaxHighlighter key={idx} language='htmlbars' style={solarizedDark}>
                        {pretty(cur, {ocd: true})}
                    </SyntaxHighlighter>          
                )
            }
        });
    }

    if (this.state.openTabs === "Headers") {
      let headerObj = this.props.responseContent.headers;

      if (!Array.isArray(headerObj) && headerObj) {
        for (let key in headerObj) {
          tabContentShownEvents.push(
            <div className={"nested-grid-2"}>
              <span className={"tertiary-title title_offset"}>{key}</span>
              <span className={"tertiary-title title_offset"}>
                {headerObj[key]}
              </span>
            </div>
          );
        }
      }
    }

    return (
      <div>
        <ul className={"tab_list"}>
          <Tab onTabSelected={this.handleTabSelect} tabName={events} />
          {/* <Tab onTabSelected={this.handleTabSelect} tabName={cookies} /> */}
          <Tab onTabSelected={this.handleTabSelect} tabName={headers} />
        </ul>
        <div className={"tab_content"}>{tabContentShownEvents}</div>
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ResponseTabs);
