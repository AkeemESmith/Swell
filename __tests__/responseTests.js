import React from 'react';
import { shallow, configure } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import ResponseContainer from '../src/client/components/containers/ResponseContainer.jsx';
import ResponseTabs from '../src/client/components/display/ResponseTabs.jsx';
import ResponseEventsDisplay from '../src/client/components/display/ResponseEventsDisplay.jsx';
import ResponseSubscriptionDisplay from '../src/client/components/display/ResponseSubscriptionDisplay.jsx';

configure({ adapter: new Adapter() });

describe('ResponseContainer', () => {
  let props;
  beforeAll(() => {
    props = {
      content: {
        request: {
          method: null
        },
      }
    }
  });

  it('should render Response Headers when selected', () => {
    const wrapper = shallow(<ResponseContainer {...props} />);
    wrapper.setState({ openTab: 'Response Headers' });
    expect(wrapper.find('ResponseHeadersDisplay')).toBeTruthy();
  });

  it('should render Response Cookies when selected', () => {
    const wrapper = shallow(<ResponseContainer {...props} />);
    wrapper.setState({ openTab: 'Response Cookies' });
    expect(wrapper.find('ResponseCookiesDisplay')).toBeTruthy();
  });

  it('should render ResponseSubscriptionDisplay for GQL Subscriptions', () => {
    const subscriptionProps = {
      content: {
        request: { method: 'SUBSCRIPTION' }
      }
    }
    const wrapper = shallow(<ResponseContainer {...subscriptionProps} />);
    expect(wrapper.find('ResponseSubscriptionDisplay')).toBeTruthy();
  });

  it('should render ResponseEventsDisplay for all other requests', () => {
    const wrapper = shallow(<ResponseContainer {...props} />);
    expect(wrapper.find('ResponseEventsDisplay')).toBeTruthy();
  });
});

describe('ResponseTabs', () => {
  it('should render three tabs', () => {
    const wrapper = shallow(<ResponseTabs />);
    expect(wrapper.find('Tab')).toHaveLength(3);
  });
});

describe('ResponseEventsDisplay', () => {
  it('if SSE, should render event rows', () => {
    const props = {
      response: {
        headers: { 'content-type': 'text/event-stream' },
        events: ['event 1', 'event 2'],
      }
    };
    const wrapper = shallow(<ResponseEventsDisplay {...props} />);
    expect(wrapper.find('SSERow')).toHaveLength(2);
  });
  it('if not SSE, should render single event', () => {
    const props = {
      response: {
        headers: null,
        events: ['event 1'],
      }
    };
    const wrapper = shallow(<ResponseEventsDisplay {...props} />);
    expect(wrapper.find('.json-response')).toHaveLength(1);
  });
});

describe('ResponseSubscriptionDisplay', () => {
  let props;
  let wrapper;
  const testURL = 'https://swell-test-graphql.herokuapp.com';
  beforeAll(() => {
    props = {
      content: {
        url: testURL,
        protocol: 'https://',
        connection: 'open',
        request: {
          body: `subscription {
            newLink {
              id
              url
              description
              postedBy {
                id
                name
                email
              }
            }
          }`,
        },
        response: {
          events: [],
        }
      },
      reqResUpdate: jest.fn(),
    };
    wrapper = renderer.create(<ResponseSubscriptionDisplay {...props} />);

  });
  xit('should initialize as listening', () => {
    expect(wrapper).toMatchSnapshot();
  });
});