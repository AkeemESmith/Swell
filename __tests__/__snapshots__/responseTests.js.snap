// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseSubscriptionDisplay should initialize as listening 1`] = `
<div
  className="tab_content-response"
>
  <div
    className="json-response"
  >
    Listening for new data
  </div>
</div>
`;
